File: ProgrammingFundamentalsExams/SampleExam2October2016/04.CubicMessages/CubicMessages.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;

class CubicMessages
{
    static void Main(string[] args)
    {
        string pattern = @"(^\d+)([A-Za-z]+)([^A-Za-z]*)$";
        string command = Console.ReadLine();

        while (command != "Over!")
        {
            Match match = Regex.Match(command, pattern);
            int count = int.Parse(Console.ReadLine());
            string message = match.Groups[2].ToString();
            StringBuilder sb = new StringBuilder();

            if (match.Success && count == message.Length)
            {
                char[] firstIndexes = match.Groups[1].ToString().ToCharArray();
                foreach (char c in firstIndexes)
                {
                    if (int.Parse(c.ToString()) <= message.Length - 1)
                    {
                        sb.Append(message[int.Parse(c.ToString())]);
                    }
                    else
                    {
                        sb.Append(" ");
                    }
                }

                char[] lastIndexes = match.Groups[3].ToString().ToCharArray();
                if (lastIndexes.Length > 0)
                {
                    foreach (char c in lastIndexes)
                    {
                        if (c >= 48 && c <= 57 && int.Parse(c.ToString()) <= message.Length - 1)
                        {
                            sb.Append(message[int.Parse(c.ToString())]);
                        }
                        else
                        {
                            sb.Append(" ");
                        }
                    }
                }
                Console.WriteLine($"{message} == {sb}");
            }            
            command = Console.ReadLine();
        }
    }
}

File: JavaAdvancedExams/JavaAdvancedExam19June2016/src/CubicMessages.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class CubicMessages {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));

        String command = reader.readLine();

        Pattern p = Pattern.compile("(\\d+)([A-Za-z]+)([^A-Za-z]*)");

        while (!command.equals("Over!")) {
            String input = command;
            int desiredLength = Integer.parseInt(reader.readLine());

            Matcher m = p.matcher(input);

            if (!m.matches() || m.group(2).length() != desiredLength) {
                command = reader.readLine();
                continue;
            }

            StringBuilder verificationCode = new StringBuilder();
            String message = m.group(2);


            for (char c : m.group(1).toCharArray()) {
                int index = Integer.parseInt(Character.toString(c));

                if (index >= 0 && index < message.length()) {
                    verificationCode.append(message.charAt(index));
                } else {
                    verificationCode.append(" ");
                }
            }

            if (m.group(3) != null) {
                for (char c : m.group(3).toCharArray()) {
                    if (Character.isDigit(c)) {
                        int index = Integer.parseInt(Character.toString(c));

                        if (index >= 0 && index < message.length()) {
                            verificationCode.append(message.charAt(index));
                        } else {
                            verificationCode.append(" ");
                        }
                    }
                }
            }

            System.out.printf("%s == %s%n", message, verificationCode);
            command = reader.readLine();
        }
    }
}
